pkg/proxy/integrations/httpparser/httpparser.go

```diff
@@ -637,3 +734,113 @@ func encodeOutgoingHttp(request []byte, clientConn, destConn net.Conn, logger *z
 	}
 	return nil
 }
+
+// ParseFinalHttp is used to parse the final http request and response and save it in a yaml file
+func ParseFinalHttp(finalReq []byte, finalResp []byte, reqTimestampMock, resTimestampcMock time.Time, ctx context.Context, logger *zap.Logger, h *hooks.Hook) error {
+	var req *http.Request
+	// converts the request message buffer to http request
+	req, err := http.ReadRequest(bufio.NewReader(bytes.NewReader(finalReq)))
+	if err != nil {
+		logger.Error("failed to parse the http request message", zap.Error(err))
+		return err
+	}
+	var reqBody []byte
+	if req.Body != nil { // Read
+		var err error
+		reqBody, err = io.ReadAll(req.Body)
+		if err != nil {
+			// TODO right way to log errors
+			logger.Error("failed to read the http request body", zap.Error(err))
+			return err
+		}
+	}
+	// converts the response message buffer to http response
+	respParsed, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(finalResp)), req)
+	if err != nil {
+		logger.Error("failed to parse the http response message", zap.Error(err))
+		return err
+	}
+	//Add the content length to the headers.
+	var respBody []byte
+	//Checking if the body of the response is empty or does not exist.
+
+	if respParsed.Body != nil { // Read
+		if respParsed.Header.Get("Content-Encoding") == "gzip" {
+			check := respParsed.Body
+			ok, reader := checkIfGzipped(check)
+			logger.Debug("The body is gzip? " + strconv.FormatBool(ok))
+			logger.Debug("", zap.Any("isGzipped", ok))
+			if ok {
+				gzipReader, err := gzip.NewReader(reader)
+				if err != nil {
+					logger.Error("failed to create a gzip reader", zap.Error(err))
+					return err
+				}
+				respParsed.Body = gzipReader
+			}
+		}
+		respBody, err = io.ReadAll(respParsed.Body)
+		if err != nil {
+			logger.Error("failed to read the the http response body", zap.Error(err))
+			return err
+		}
+		logger.Debug("This is the response body: " + string(respBody))
+		//Set the content length to the headers.
+		respParsed.Header.Set("Content-Length", strconv.Itoa(len(respBody)))
+	}
+	// store the request and responses as mocks
+	meta := map[string]string{
+		"name":      "Http",
+		"type":      models.HttpClient,
+		"operation": req.Method,
+	}
+	passthroughHost := false
+	for _, host := range models.PassThroughHosts {
+		if req.Host == host {
+			passthroughHost = true
+		}
+	}
+	if !passthroughHost {
+		err = h.AppendMocks(&models.Mock{
```

Comment:
This should be in a go routine else parser will get struck here sometimes.

Reply:
Okay will add go routine here.